### PR TITLE
Check that the SIMPLE card is present

### LINKED ISSUE #3941: More descriptive/accurate error when opening non-FITS files
I tried to open a FITS file that happened to be bzip2 compressed (e.g. `frame-g-001000-1-0027.fits.bz2`). This is the error I got:

```
Traceback (most recent call last):
  File "./populate_frame_polygons.py", line 102, in <module>
    points = get_polygon_points(filepath=filepath, hdu=0)
  File "./populate_frame_polygons.py", line 44, in get_polygon_points
    image_hdu = fits.open(filepath)[hdu]
  File "/uufs/chpc.utah.edu/sys/pkg/python/2.7.3_rhel6/lib/python2.7/site-packages/astropy-1.0.2-py2.7-linux-x86_64.egg/astropy/io/fits/hdu/hdulist.py", line 129, in fitsopen
    return HDUList.fromfile(name, mode, memmap, save_backup, cache, **kwargs)
  File "/uufs/chpc.utah.edu/sys/pkg/python/2.7.3_rhel6/lib/python2.7/site-packages/astropy-1.0.2-py2.7-linux-x86_64.egg/astropy/io/fits/hdu/hdulist.py", line 271, in fromfile
    save_backup=save_backup, cache=cache, **kwargs)
  File "/uufs/chpc.utah.edu/sys/pkg/python/2.7.3_rhel6/lib/python2.7/site-packages/astropy-1.0.2-py2.7-linux-x86_64.egg/astropy/io/fits/hdu/hdulist.py", line 829, in _readfrom
    hdu = _BaseHDU.readfrom(ffo, **kwargs)
  File "/uufs/chpc.utah.edu/sys/pkg/python/2.7.3_rhel6/lib/python2.7/site-packages/astropy-1.0.2-py2.7-linux-x86_64.egg/astropy/io/fits/hdu/base.py", line 378, in readfrom
    **kwargs)
  File "/uufs/chpc.utah.edu/sys/pkg/python/2.7.3_rhel6/lib/python2.7/site-packages/astropy-1.0.2-py2.7-linux-x86_64.egg/astropy/io/fits/hdu/base.py", line 438, in _readfrom_internal
    header = Header.fromfile(data, endcard=not ignore_missing_end)
  File "/uufs/chpc.utah.edu/sys/pkg/python/2.7.3_rhel6/lib/python2.7/site-packages/astropy-1.0.2-py2.7-linux-x86_64.egg/astropy/io/fits/header.py", line 460, in fromfile
    padding)[1]
  File "/uufs/chpc.utah.edu/sys/pkg/python/2.7.3_rhel6/lib/python2.7/site-packages/astropy-1.0.2-py2.7-linux-x86_64.egg/astropy/io/fits/header.py", line 529, in _from_blocks
    raise IOError('Header missing END card.')
IOError: Header missing END card.
```

The error incorrectly points out that the header is missing an END card; rather, the problem is that the file being opened was compressed. It should be easy to check for obvious problems like this.

Alternatively (preferably), it would be great if io.fits could just go ahead and open the file, e.g.:

```
import bz2
if filename.endswith(".bz2"):
    decompressed_file = bz2.BZ2File(filename)
# continue with opening that file...
```

### LINKED ISSUE #7569: fits.open should raise useful exception when it's given a non FITS file
Dear all,

Using the following code:

```
from astroquery.esasky import ESASky
ESASky.list_maps()
maps = ESASky.query_object_maps('13 29 52.7 +47 11 43')
maps_data = ESASky.get_maps(maps,'all', download_dir = '/Users/astro_user/Downloads')

I get a long list of warning for HST only (the other mission work perfectly), like:

> WARNING: The following header keyword is invalid or follows an unrecognized non-standard convention:
ustar00 [astropy.io.fits.card]

Which finally ends with generating the following error messages and stops the download process:

> Downloading Observation ID: n48r020n0 from http://archives.esac.esa.int/ehst-sl-server/servlet/data-action?RETRIEVAL_TYPE=PRODUCT&OBSERVATION_ID=n48r020n0 Traceback (most recent call last):

  File "<ipython-input-2-d0bbe0f5ca3a>", line 1, in <module>
    maps_data = ESASky.get_maps(maps,'HST', download_dir = '/Users/Bastian/Downloads')

  File "/anaconda3/lib/python3.6/site-packages/astroquery/esasky/core.py", line 374, in get_maps
    cache))

  File "/anaconda3/lib/python3.6/site-packages/astroquery/esasky/core.py", line 548, in _get_maps_for_mission
    maps.append(fits.open(file_name))

  File "/anaconda3/lib/python3.6/site-packages/astropy/io/fits/hdu/hdulist.py", line 164, in fitsopen
    lazy_load_hdus, **kwargs)

  File "/anaconda3/lib/python3.6/site-packages/astropy/io/fits/hdu/hdulist.py", line 400, in fromfile
    lazy_load_hdus=lazy_load_hdus, **kwargs)

  File "/anaconda3/lib/python3.6/site-packages/astropy/io/fits/hdu/hdulist.py", line 1012, in _readfrom
    read_one = hdulist._read_next_hdu()

  File "/anaconda3/lib/python3.6/site-packages/astropy/io/fits/hdu/hdulist.py", line 1087, in _read_next_hdu
    hdu = _BaseHDU.readfrom(fileobj, **kwargs)

  File "/anaconda3/lib/python3.6/site-packages/astropy/io/fits/hdu/base.py", line 328, in readfrom
    **kwargs)

  File "/anaconda3/lib/python3.6/site-packages/astropy/io/fits/hdu/base.py", line 393, in _readfrom_internal
    header = Header.fromfile(data, endcard=not ignore_missing_end)

  File "/anaconda3/lib/python3.6/site-packages/astropy/io/fits/header.py", line 447, in fromfile
    padding)[1]

  File "/anaconda3/lib/python3.6/site-packages/astropy/io/fits/header.py", line 516, in _from_blocks
    raise OSError('Header missing END card.')

OSError: Header missing END card.

Many thanks for your help.

Best,

Bastian

### PR DESCRIPTION
**Pull request by saimn**
Fixes #3941, fixes #7569.
With this `fits.open` will check that a `SIMPLE` card is present at the beginning of file, which allows to directly raise a better error when opening a file that is not FITS. The `ignore_missing_simple` option can be used to deactivate this behavior.
@dhomeier - requesting your review since you have been interested in FITS lately ;). cc @embray if you want to have a look.

### CODE REVIEW

**Comment by embray** [comment]
Do we really need the `ignore_missing_simple` option?  As I see it this is just one more verification, and doesn't need a dedicated argument.

I still have an interface in mind for how I would like to better manage FITS standard verification (whether I will ever have time to try to implement it is another story, admittedly).

In other words, I would rather see this implemented within a verification framework (which currently does not exist in a consistent manner).  In the meantime, instead of an [`OSError`](https://docs.python.org/3/library/exceptions.html#OSError) I think this should raise a `VerifyError`.

File: astropy/io/fits/hdu/hdulist.py [hunk 1]
```diff
@@ -1039,7 +1047,8 @@ def filename(self):
 
     @classmethod
     def _readfrom(cls, fileobj=None, data=None, mode=None, memmap=None,
-                  cache=True, lazy_load_hdus=True, **kwargs):
+                  cache=True, lazy_load_hdus=True, ignore_missing_simple=False,
```
Comment (embray) [inline astropy/io/fits/hdu/hdulist.py:1050]: Per my previous comment I'm not sure we even need an option for this; absent a more generic "verification" framework I don't think there's ever any reason not to check this and not to consider it an error if the `SIMPLE` keyword is missing.

File: astropy/io/fits/hdu/hdulist.py [hunk 2]
```diff
@@ -1065,6 +1074,29 @@ def _readfrom(cls, fileobj=None, data=None, mode=None, memmap=None,
...
+            pos = hdulist._file.tell()
+            raise_error = False
+
+            try:
+                simple = hdulist._file.read(30).decode('ascii')
+            except UnicodeDecodeError:
+                raise_error = True
+            else:
+                if simple != 'SIMPLE  =                    T':
+                    if simple == 'SIMPLE  =                    F':
```
Comment (embray) [inline astropy/io/fits/hdu/hdulist.py:1087]: Instead of calling `decode('ascii')` above could it not just compare directly to `FITS_SIGNATURE[:-1]` plus ensure `simple[-1] in (b'T', b'F')`?

**Comment by embray** [comment]
TL;DR +1 to adding this (I believe it has been requested in the past), but I think we can make it even simpler (no pun intended :)

File: astropy/io/fits/hdu/hdulist.py [hunk 3]
```diff
@@ -1065,6 +1074,29 @@ def _readfrom(cls, fileobj=None, data=None, mode=None, memmap=None,
             # fromstring case; the data type of ``data`` will be checked in the
             # _BaseHDU.fromstring call.
 
+        if hdulist._file and hdulist._file.readonly:
```
[the remark this answers is not in the document]
Comment (dhomeier) [reply astropy/io/fits/hdu/hdulist.py:1077]: An idea how to implement this? But important point; this should be tested on a non-seekable file object.

**Comment by dhomeier** [comment]
I very much agree with @embray's overall approval and suggestions.
The main open issue seems to be whether we want a specific new keyword `ignore_missing_simple` for this; this is only one specific validation, but the only other one that can be controlled with its own kwarg I found is `ignore_missing_end`. At the very least I think the two should follow each other in the docstring. `ignore_missing_end` then is silently (in `**kwargs`) handed through to `_BaseHDU._readfrom_internal`, so an obvious question is whether both checks could not be done consistently in one place. But that might not be feasible as the full header has to be read in before detecting the missing 'END', and I see the value of having an immediate validity check on the first 80 bytes.

I'd say the underlying policy question is how differently we want to treat different levels of standard violations. Per the current standard (§ 4.4.1.1) a minimum requirement is actually having
```
SIMPLE = T
BITPIX = i
NAXIS = n
NAXIS1
...
NAXISn
...
END
```
and per this PR various violations of this standard would be treated as

1. No `SIMPLE` card: "not a FITS file"

2. `SIMPLE != T`: `_NonstandardHDU`

3. No `END` card: `OSError`

The latter two should probably be treated as the same level, and as pointed out in https://github.com/astropy/astropy/pull/10895#issuecomment-712298007 as `VerifyError` pending a consistent validation framework. I guess violations of the `NAXIS` standards may trigger validation errors later in the parsing.

Then it remains open if the first case should be treated as a more serious violation (so perhaps we could still stick with two kwargs, one to allow general nonstandard HDUs including `SIMPLE = F`, and one for the entirely missing `SIMPLE`. From what I gleaned from the discussion on #3941, the latter unfortunately still occur quite frequently.

Final question: how does this go along with the still existing https://github.com/astropy/astropy/blob/2fa4824ec36ff7bb68f41c5e271d5c3f0163e34b/astropy/io/fits/connect.py#L53-L57
e.g., will
`fits.open(open(filename, mode='rb'), ignore_missing_simple=True)`
work as well?

**Comment by saimn** [comment]
Re `ignore_missing_simple`: it's simply the same idea as `ignore_missing_end`, `io.fits` is able to read invalid files and has done so for a long time, and I'm pretty sure there will be files without the `SIMPLE` card. So basically it's just a way to allow users reading their broken files to fix them. And we just got an example today where `io.fits` is able to read a really broken file : #10905

File: astropy/io/fits/hdu/hdulist.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (saimn) [reply astropy/io/fits/hdu/hdulist.py:1077]: Hmm, I started this a long time ago so I'm not sure exactly why, and `.readonly` is set for the `('readonly', 'copyonwrite', 'denywrite')` modes. I guess it should work also work for `append` and `update`, but we should exclude `ostream`.

File: astropy/io/fits/hdu/hdulist.py (hunk 2 above)
Comment (saimn) [reply astropy/io/fits/hdu/hdulist.py:1087]: Yeah indeed, I think I went that way because of one of the unit test using some unicode characters, but I agree it can be simplified!

**Comment by saimn** [comment]
> so an obvious question is whether both checks could not be done consistently in one place. But that might not be feasible as the full header has to be read in before detecting the missing 'END', and I see the value of having an immediate validity check on the first 80 bytes.

Yep, I guess you got the answer ;). 

> I guess violations of the NAXIS standards may trigger validation errors later in the parsing.

Yes this one is handled later by the verification system, where we also fix some common issues like using strings instead of integers (for NAXIS).

> Final question: how does this go along with the still existing 

The code you are pointing is only to detect FITS file for the unified IO (i.e. with `Table.read`), when using a file object.

File: astropy/io/fits/tests/test_hdulist.py [hunk 4]
```diff
@@ -938,16 +1004,14 @@ def test_proper_error_raised_on_non_fits_file_with_unicode(self):
...
             f.write('Ce\xe7i ne marche pas')
 
         # This should raise an OSError because there is no end card.
-        with pytest.raises(OSError):
-            with pytest.warns(AstropyUserWarning, match=r'non-ASCII characters '
-                              r'are present in the FITS file header'):
-                fits.open(filename)
+        with pytest.raises(OSError, match='No SIMPLE card found, this file '
+                           'does not appear to be a valid FITS file'):
+            fits.open(filename)
```
Comment (dhomeier) [inline astropy/io/fits/tests/test_hdulist.py:1014]: With
```
fits.open(filename, ignore_missing_simple=True)
```
this should still raise the `AstropyUserWarning` once it encounters the non-ASCII char, right? Could this case be tested additionally?

**Review by dhomeier — CHANGES_REQUESTED** [review]
Thanks for the updates, apart from the comments on the last test this looks good to go!

File: astropy/io/fits/tests/test_hdulist.py (hunk 4 above)
Comment (saimn) [reply astropy/io/fits/tests/test_hdulist.py:1014]: I don't think so, since this is an invalid file the error must be raised instead of the warning.
I guess one question is to know is there is another test with non-ASCII char in a valid header/file, I can check that later.

File: astropy/io/fits/hdu/hdulist.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (saimn) [reply astropy/io/fits/hdu/hdulist.py]: With the recent changes the check is also done for `append` and `update`, only `ostream` is excluded.

**Comment by saimn** [comment]
@embray @dhomeier : So, I guess the main point remaining is whether to keep `ignore_missing_simple` or not. I'm happy to remove it if you think it's not needed. I added it because I fear the variety of invalid FITS files lying around, but I guess we can also tell people to fix their files if they miss the SIMPLE card.

File: astropy/io/fits/tests/test_hdulist.py (hunk 4 above)
Comment (dhomeier) [reply astropy/io/fits/tests/test_hdulist.py:1014]: Yes, would be good. My point was, `ignore_missing_simple=True` is supposed to allow people to open otherwise valid files that are _only_ missing the SIMPLE card, but an otherwise broken should still raise an exception (which would rather not be the one about missing SIMPLE). But I don't know the details of #5594 and if it still requires a dedicated test.

**Comment by dhomeier** [comment]
> @embray @dhomeier : So, I guess the main point remaining is whether to keep `ignore_missing_simple` or not. I'm happy to remove it if you think it's not needed. I added it because I fear the variety of invalid FITS files lying around, but I guess we can also tell people to fix their files if they miss the SIMPLE card.

Not sure how feasible that is if there are, as it seems, lots of them even coming out of major pipelines, where people have little control of the format they get. Can any of our FITS utilities fix it without requiring an expert user? Then I'd vote for that option, but otherwise I am fine with offering the option in `fitsopen`. I think the debate was more on how many individual override options we need, but since this a fairly fundamental one and implemented very early on, it seems legit to give it its own option for now.

The non-ascii test issue should still be resolved.

**Comment by saimn** [comment]
I checked and the non-ASCII case is tested by `test_header::test_invalid_characters`.
Yes, that's the point of having this parameter. To offer the possibility to read a file missing `SIMPLE` if it can be read, because then reading and writing it with Astropy would allow to fix it.

**Comment by dhomeier** [comment]
> Yes, that's the point of having this parameter. To offer the possibility to read a file missing `SIMPLE` if it can be read, because then reading and writing it with Astropy would allow to fix it.

Yes, `fitscheck` currently cannot fix this (it currently cannot even handle non-standard headers like `SIMPLE = F` or missing `END` card); basically to deal with missing `SIMPLE` would most likely require a new option for opening the file with `ignore_missing_simple`, so `fitsopen` would need that option anyway.

+1

File: astropy/io/fits/hdu/hdulist.py [hunk 5]
```diff
@@ -94,6 +98,12 @@ def fitsopen(name, mode='readonly', memmap=None, save_backup=False,
         Do not issue an exception when opening a file that is missing an
         ``END`` card in the last header. Default is `False`.
 
+    ignore_missing_simple : bool, optional
+        Do not issue an exception when the SIMPLE keyword is missing.
```
Comment (dhomeier) [inline astropy/io/fits/hdu/hdulist.py:102]:
```suggestion
        Do not raise an exception when opening a file that is missing an
        ``END`` card in the last header. Default is `False`.

    ignore_missing_simple : bool, optional
        Do not raise an exception when the SIMPLE keyword is missing.
```
Nitpick: I think `raise` is the standard terminology; not sure if there are official recommendations, but to be that verb makes it clearer that program execution will be interrupted.

**Comment by dhomeier** [comment]
It seems there are still issues with the treatment of _invalid_ headers:
With `SIMPLE = F` in the primary header the first HDU is opened as a `_NonstandardHDU` but then parsing apparently stops, reading only the first HDU. And the calling code does not handle this well, e.g. `fitscheck` is failing on such a file with
```
OK '_broken.fits'
EXCEPTION '_broken.fits' .. '_NonstandardHDU' object has no attribute 'add_checksum'
1 errors
```
This behaviour is not affected by this PR and rather belongs in a separate issue, but probably confirms the need for a more consistent verification framework @embray pointed out.

**Comment by embray** [comment]
>  and I'm pretty sure there will be files without the SIMPLE card

I'm not so sure.  Of all the crazy, bogus, busted FITS files I've seen, I have never seen one like that.  It should just not be treated as a valid FITS file.

**Comment by embray** [comment]
Also, if I had the opportunity to overhaul the FITS verification framework I would also remove the `ignore_missing_end` keyword argument (after deprecation of course).  I don't remember why it was added exactly, except that long ago somebody needed it because there were some such observatory that was producing such invalid files.  But I've never heard of a missing `SIMPLE` keyword except for files that are not FITS files at all.  I'd prefer to not add another keyword argument that isn't known to be actually  needed until and unless a specific reason for it is given for a real world file.  Otherwise it's just additional stuff that might be removed later.

**Comment by dhomeier** [comment]
> > and I'm pretty sure there will be files without the SIMPLE card
> 
> I'm not so sure. Of all the crazy, bogus, busted FITS files I've seen, I have never seen one like that. It should just not be treated as a valid FITS file.

Maybe I misinterpreted some comments in the earlier discussion like https://github.com/astropy/astropy/issues/3941#issuecomment-121329240 when I concluded that they were not even uncommon.
If there is consensus that there is no necessity to handle such broken files at all, of course the options could be removed/deprecated. If some files so seriously broken do exist after all, the main problem to me is that I don't see much of a middle way between

1. Offering a fairly convenient option to open them almost like a valid FITS file and

3. Telling users to ask whoever has originally created the files to provide a valid version, or write some low-level cfitsio wrapper themselves to fix them "manually".

I think some possibilities to make this support slightly less "official" could be

1. Stick to only a single option like `ignore_fits_violations` to make it very clear those are non-standard files.

2. Offer it only at a lower level like `HDUList.fromfile` or `HDUList._readfrom` – not sure if this is good policy?

**Comment by embray** [comment]
@dhomeier You raise some good points.  My thinking was that the string `SIMPLE  =                    T` is the FITS magic string identifying a file as a FITS file.  But nowhere does the FITS Standard make this explicit--this is only specified in RFC 4047 which is not incorporated into the FITS Standard.  The Standard does say that the primary HDU must begin with `SIMPLE  = T` in order to conform to the FITS Standard, but if it's missing that does not mean the file is not still in some sense a non-standard-compliant FITS file (there are plenty of other pseudo-FITS files out there that also don't perfectly conform to the standard, yet we still read them).

It's also a standard violation that can easily be "fixed", as we do with other violations, so long as the rest of the file "looks like" a FITS file (e.g., the first 80 bytes at least parses as a valid header card of some kind).

My other concern about adding an explicit option for this is I would rather see it handled in the context of a more general validation framework that also has the ability to fix simple errors like this as long as the rest of the file is readable.  Years ago I started on that project but never completed it, and I think it would have to be resurrected from the ground up.  It consisted of two layers:

* Checking the basic syntax of FITS files, were all possible syntax violations were classified explicitly, and with the ability to register custom fixers for specific violations (including many built-in fixers for violations where the fix is fairly unambiguous).  This of course already exists on some level in astropy.io.fits, but it's scattered about the code in an ad-hoc manner that's hard to reason about.
* A higher-level "FITS Schema" that could be used to validate the structure of otherwise syntactically correct headers.

TL;DR, absent time to resurrect and complete that work though, I agree it would be unfair to say "no new arguments can be added to `fits.open()` because it might be superseded by future work that may or may not ever be completed.  So I rescind my objection to adding the argument for now.  It can simply be deprecated along with `ignore_missing_end` if the verification framework is ever overhauled.

File: astropy/io/fits/hdu/hdulist.py (hunk 5 above)
Comment (saimn) [reply astropy/io/fits/hdu/hdulist.py]: Indeed, `raise` seems better. I did the change locally since I needed to rebase and move the changelog entry, to avoid running the CI twice.

**Comment by saimn** [comment]
(Rebased and moved the changelog entry to 4.3).

I would have been happy to remove this option, but it seems we now agree on keeping it, and we could indeed deprecate/remove it if one day there is a better verification system. So I think it's good to go now ?

**Review by dhomeier — APPROVED** [review]

**Comment by maartenbreddels** [comment]
This PR triggered a CI failure in Vaex, fix in https://github.com/vaexio/vaex/pull/1490

Our header had:
```
b'SIMPLE  = T                   '
```

Our header/value was aligned to the right with spaces, I couldn't find that in the spec/rfc that it had to be that way, but in case it does not, this check might have been too strict.

**Comment by saimn** [comment]
SIMPLE is a mandatory keyword so it's value must be in Byte 30: (from the FITS standard, https://fits.gsfc.nasa.gov/fits_standard.html)

image

image

RFC 4047 also specifies the magic number : (https://datatracker.ietf.org/doc/html/rfc4047)

image

**Comment by maartenbreddels** [comment]
Thanks, then it was simply an error on our side!

**Comment by saimn** [comment]
It was actually a good question, finding the justification was not so easy ;)

**Comment by maartenbreddels** [comment]
Yeah, good to have this sorted out, thanks for that!

**Comment by kYwzor** [comment]
I have a similar issue where the first 30 bytes of some external data are
```
b'SIMPLE  = T /Tape is in Fits f'
```
instead of the expected
```
b'SIMPLE  =                    T'
```

so I got `OSError: No SIMPLE card found, this file does not appear to be a valid FITS file`, which had me scratching my head for a while, as I was sure the file had a SIMPLE card and Astropy used to recognise these files in previous versions. In fact, if we use the new `ignore_missing_simple` option, Astropy 4.3 will still eventually figure out that the file does have a SIMPLE card (due to the techniques it uses to interpret non-standard files).

Don't get me wrong I completely agree that the file I have is non-standard, but it seems to me that this check is overly strict if our goal is just to be sure that we're dealing with a FITS file. Keeping in mind this philosophy:

https://github.com/astropy/astropy/blob/c6cafcb763c9a2cf6cbbc1989d7fbb6dbdb6932f/docs/io/fits/usage/verification.rst#L8-L15

I believe we could make a much better attempt at finding a SIMPLE card before giving up. Instead of:
https://github.com/astropy/astropy/blob/c6cafcb763c9a2cf6cbbc1989d7fbb6dbdb6932f/astropy/io/fits/hdu/hdulist.py#L1084-L1085
Why not something like:
`match_sig = re.match(rb"SIMPLE *= *[T|F]", simple)`

**Comment by maartenbreddels** [comment]
I like that, a warning would be nice though so that people are aware the fits image is not created according to the standard.

**Comment by kYwzor** [comment]
@saimn @embray Could you please give your opinion on my comment above? This breaks backwards compatibility in my use case, 
Would you prefer that I create a separate issue? Thank you.

**Comment by saimn** [comment]
Sorry for the delay, last 2 weeks have been quite busy.
I think I would prefer to keep the check as is, at least for now. As described above this card is well defined and is the only way to identify a FITS file. `io.fits` is more flexible for the other cards, and tries to fix them if possible, but the case of `SIMPLE` is a bit different given its role to identify a FITS file.
Depending on the number of complaint we get we could revisit and issue a warning if SIMPLE is there but not valid, but I think for now `ignore_missing_simple` is a good enough workaround (and let's hope this encourages people to fix their broken files or report to the producer!).

**Comment by kYwzor** [comment]
No worries, take your time. Correct me if I'm wrong, but from what I can tell the main goals here were that, when `io.fits` faces a file that it _cannot_ interpret as FITS, it should 1) stop the attempt earlier and 2) raise a better/clearer error. My problem with the current change is that it triggers the error on some non-standard files that `io.fits` _can_ interpret as FITS. The change that I'm suggesting would still achieve those two main goals, while being a bit more lax with the verification. I think this would be more in line with the stated philosophy of being "tolerant with input and strict with output" and not giving up as soon as the file is deemed non-standard. I believe issuing a warning would make it more similar to the behaviour of other standard violations, e.g. null bytes being used for padding. To be clear, the change I proposed would only affect cases where `SIMPLE` is present but the spacing is non-standard, and it would still raise an error if `SIMPLE` is non-existent.

Maybe I understood it incorrectly and the goal here is specifically to discourage the usage of non-standard `SIMPLE` cards? I can see the benefits there, but I would guess that most people that have to deal with non-standard files aren't the ones actually producing those files... and if they are, a warning can still get the message across.

Either way, even if this behaviour isn't changed, could we at least include a suggestion to use `ignore_missing_simple` in the error message? This could help reduce the confusion of why a file suddenly cannot be read after an update. This was especially confusing in my case where the SIMPLE keywords _was_ there, just in a non-standard way. Or at the very least the error message could be different in that case.

**Comment by melanieclarke** [comment]
This is a significant break in backwards compatibility for our code that processes archival data that cannot be easily changed.  We will have to update hundreds of lines of code -- every fits.open we have in 11 different packages -- in order to set ignore_missing_simple=True.  Would it at least be possible to make True the default value to preserve handling for files that can be read as FITS?

**Comment by pllim** [comment]
It is not impossible to add a config item to control this, though it does add to code complexity... @melanieclarke , can you please open a new issue and link to this one? Thanks.
